libtenzir/builtins/contexts/geoip.cpp

```diff
@@ -0,0 +1,501 @@
+            r[sub_record_key] = sub_r.begin()->second;
+          } else if (not sub_r.empty()) {
+            r[sub_record_key] = sub_r;
+          }
+        }
+        break;
+      }
+      case MMDB_DATA_TYPE_ARRAY: {
+        auto l = list{};
+        auto size = entry_data_list->entry_data.data_size;
+        auto sub_r = record{};
+        for (entry_data_list = entry_data_list->next; size && entry_data_list;
+             size--) {
+          entry_data_list = entry_data_list_to_list(entry_data_list, status, l);
+          if (*status != MMDB_SUCCESS) {
+            return entry_data_list;
+          }
+        }
+        r[key] = l;
+        break;
+      }
+      case MMDB_DATA_TYPE_UTF8_STRING: {
+        auto str = std::string{entry_data_list->entry_data.utf8_string,
+                               entry_data_list->entry_data.data_size};
+        r[key] = str;
+        entry_data_list = entry_data_list->next;
+        break;
+      }
+      case MMDB_DATA_TYPE_BYTES: {
+        auto bytes = blob{
+          reinterpret_cast<const std::byte*>(entry_data_list->entry_data.bytes),
+          entry_data_list->entry_data.data_size};
+        r[key] = std::move(bytes);
+        entry_data_list = entry_data_list->next;
+        break;
+      }
+      case MMDB_DATA_TYPE_DOUBLE:
+        r[key] = entry_data_list->entry_data.double_value;
+        entry_data_list = entry_data_list->next;
+        break;
+      case MMDB_DATA_TYPE_FLOAT:
+        r[key] = entry_data_list->entry_data.float_value;
+        entry_data_list = entry_data_list->next;
+        break;
+      case MMDB_DATA_TYPE_UINT16:
+        r[key] = entry_data_list->entry_data.uint16;
+        entry_data_list = entry_data_list->next;
+        break;
+      case MMDB_DATA_TYPE_UINT32:
+        r[key] = entry_data_list->entry_data.uint32;
+        entry_data_list = entry_data_list->next;
+        break;
+      case MMDB_DATA_TYPE_BOOLEAN:
+        r[key] = entry_data_list->entry_data.boolean;
+        entry_data_list = entry_data_list->next;
+        break;
+      case MMDB_DATA_TYPE_UINT64:
+        r[key] = entry_data_list->entry_data.uint64;
+        entry_data_list = entry_data_list->next;
+        break;
+      case MMDB_DATA_TYPE_UINT128:
+        r[key] = cast_128_bit_unsigned_to_64_bit(
+          entry_data_list->entry_data.uint128);
+        break;
+      case MMDB_DATA_TYPE_INT32:
+        r[key] = int64_t{entry_data_list->entry_data.int32};
+        entry_data_list = entry_data_list->next;
+        break;
+      default:
+        *status = MMDB_INVALID_DATA_ERROR;
+        return entry_data_list;
+    }
+    *status = MMDB_SUCCESS;
+    return entry_data_list;
+  }
+
+  /// Emits context information for every event in `slice` in order.
+  auto apply(table_slice slice, context::parameter_map parameters) const
+    -> caf::expected<std::vector<typed_array>> override {
+    auto mmdb = MMDB_s{};
+    auto status = MMDB_open(db_path_.c_str(), MMDB_MODE_MMAP, &mmdb);
```

Comment:
I attempted a little experiment. I ran `enrich` on 5000 Suricata entries ( `from ... read | head | enrich`), the first time with the code as it is now, the second time with a context instance-wide DB field that does not get created for every `apply`, only  for every update.

First code (create DB for every `apply`):

```
operator #1 (load)
  total: 21.64s
  scheduled: 11.6s (53.61%)
  processing: 1.39s (6.41%)
  runs: 1975492 (100.00% processing / 100.00% input / 0.00% output)
  outbound:
    bytes: 8388608 at a rate of 387697.45/s
    batches: 8 (1048576.00 bytes/batch)
operator #2 (read)
  total: 21.64s
  scheduled: 21.42s (98.99%)
  processing: 19.95s (92.21%)
  runs: 9043 (99.93% processing / 30.82% input / 99.93% output)
  inbound:
    bytes: 8388608 at a rate of 387718.46/s
    batches: 8 (1048576.00 bytes/batch)
  outbound:
    events: 5423 at a rate of 250.65/s
    bytes: 1847917 at a rate of 85410.06/s (estimate)
    batches: 2777 (1.95 events/batch)
operator #3 (slice)
  total: 25.91s
  scheduled: 13.59s (52.43%)
  processing: 1.88s (7.25%)
  runs: 1778735 (100.00% processing / 0.15% input / 100.00% output)
  inbound:
    events: 5423 at a rate of 209.30/s
    bytes: 1847917 at a rate of 71319.94/s (estimate)
    batches: 2777 (1.95 events/batch)
  outbound:
    events: 5000 at a rate of 192.97/s
    bytes: 1690594 at a rate of 65248.10/s (estimate)
    batches: 2651 (1.89 events/batch)
operator #4 (context)
  total: 56.14s
  scheduled: 22.62s (40.30%)
  processing: 15.96s (28.44%)
  runs: 449848 (100.00% processing / 1.19% input / 99.99% output)
  inbound:
    events: 5000 at a rate of 89.06/s
    bytes: 1690594 at a rate of 30114.47/s (estimate)
    batches: 2651 (1.89 events/batch)
  outbound:
    events: 5000 at a rate of 89.06/s
    bytes: 4245882 at a rate of 75631.69/s (estimate)
    batches: 2719 (1.84 events/batch)
operator #5 (write)
  total: 59.25s
  scheduled: 48.0s (81.01%)
  processing: 42.75s (72.15%)
  runs: 567518 (100.00% processing / 0.48% input / 99.99% output)
  inbound:
    events: 5000 at a rate of 84.39/s
    bytes: 4245882 at a rate of 71661.70/s (estimate)
    batches: 2719 (1.84 events/batch)
  outbound:
    bytes: 11911266 at a rate of 201037.52/s
    batches: 2721 (4377.53 bytes/batch)
operator #6 (save)
  total: 59.26s
  scheduled: 32.13s (54.22%)
  processing: 4.43s (7.48%)
  runs: 6305119 (100.00% processing / 0.04% input / 100.00% output)
  inbound:
    bytes: 11911266 at a rate of 200987.91/s
    batches: 2721 (4377.53 bytes/batch)
```

Second code ( DB context-wide):

```
operator #1 (load)
  total: 22.33s
  scheduled: 11.95s (53.52%)
  processing: 1.32s (5.93%)
  runs: 2074747 (100.00% processing / 100.00% input / 0.00% output)
  outbound:
    bytes: 8388608 at a rate of 375660.51/s
    batches: 8 (1048576.00 bytes/batch)
operator #2 (read)
  total: 22.32s
  scheduled: 21.96s (98.39%)
  processing: 20.46s (91.66%)
  runs: 9473 (99.94% processing / 30.20% input / 99.94% output)
  inbound:
    bytes: 8388608 at a rate of 375863.96/s
    batches: 8 (1048576.00 bytes/batch)
  outbound:
    events: 5495 at a rate of 246.21/s
    bytes: 1873332 at a rate of 83937.40/s (estimate)
    batches: 2789 (1.97 events/batch)
operator #3 (slice)
  total: 26.94s
  scheduled: 13.93s (51.71%)
  processing: 2.0s (7.41%)
  runs: 1790937 (100.00% processing / 0.15% input / 100.00% output)
  inbound:
    events: 5495 at a rate of 203.95/s
    bytes: 1873332 at a rate of 69531.36/s (estimate)
    batches: 2789 (1.97 events/batch)
  outbound:
    events: 5000 at a rate of 185.58/s
    bytes: 1690594 at a rate of 62748.78/s (estimate)
    batches: 2651 (1.89 events/batch)
operator #4 (context)
  total: 45.93s
  scheduled: 18.33s (39.92%)
  processing: 13.19s (28.71%)
  runs: 399076 (100.00% processing / 1.35% input / 99.99% output)
  inbound:
    events: 5000 at a rate of 108.86/s
    bytes: 1690594 at a rate of 36809.27/s (estimate)
    batches: 2651 (1.89 events/batch)
  outbound:
    events: 5000 at a rate of 108.86/s
    bytes: 4245882 at a rate of 92445.50/s (estimate)
    batches: 2719 (1.84 events/batch)
operator #5 (write)
  total: 49.1s
  scheduled: 40.75s (83.01%)
  processing: 36.57s (74.49%)
  runs: 532714 (100.00% processing / 0.51% input / 99.99% output)
  inbound:
    events: 5000 at a rate of 101.84/s
    bytes: 4245882 at a rate of 86481.68/s (estimate)
    batches: 2719 (1.84 events/batch)
  outbound:
    bytes: 11911266 at a rate of 242613.02/s
    batches: 2723 (4374.32 bytes/batch)
operator #6 (save)
  total: 49.11s
  scheduled: 26.46s (53.88%)
  processing: 3.83s (7.79%)
  runs: 5152169 (100.00% processing / 0.05% input / 100.00% output)
  inbound:
    bytes: 11911266 at a rate of 242518.78/s
    batches: 2723 (4374.32 bytes/batch)
```

10 seconds time saving. I should maybe implement a context-wide DB instance like that.

While measuring I also discovered a small bug that I fixed in https://github.com/tenzir/tenzir-plugins/pull/142 - can you also check & approve that PR?

Reply:
I also remember talking in Slack about this. If we keep the DB alive with the context, it would need to react to the file changes (update DB when file watcher detects an update to the file). Do we have a file watching mechanism somewhere?